lib/ansible/modules/identity/keycloak/keycloak_client.py

```diff
@@ -0,0 +1,594 @@
+    if 'attributes' in result:
+        if 'saml.signing.private.key' in result['attributes']:
+            result['attributes']['saml.signing.private.key'] = 'no_log'
+    return result
+
+
+def main():
+    """
+    Module execution
+
+    :return:
+    """
+    argument_spec = dict(
+        state=dict(default='present', choices=['present', 'absent']),
+        auth_keycloak_url=dict(type='str', aliases=['url'], required=True),
+        auth_client_id=dict(type='str', default='admin-cli'),
+        auth_realm=dict(type='str', required=True),
+        auth_client_secret=dict(type='str', default=None),
+        auth_username=dict(type='str', aliases=['username'], required=True),
+        auth_password=dict(type='str', aliases=['password'], required=True, no_log=True),
+        validate_certs=dict(type='str', default=True),
+        realm=dict(type='str', default='master'),
+
+        id=dict(type='str'),
+        client_id=dict(type='str'),
+        name=dict(type='str'),
+        description=dict(type='str'),
+        root_url=dict(type='str'),
+        admin_url=dict(type='str'),
+        base_url=dict(type='str'),
+        surrogate_auth_required=dict(type='bool'),
+        enabled=dict(type='bool'),
+        client_authenticator_type=dict(type='str', choices=['client-secret', 'client-jwt']),
+        secret=dict(type='str', no_log=True),
+        registration_access_token=dict(type='str'),
+        default_roles=dict(type='list'),
+        redirect_uris=dict(type='list'),
+        web_origins=dict(type='list'),
+        not_before=dict(type='int'),
+        bearer_only=dict(type='bool'),
+        consent_required=dict(type='bool'),
+        standard_flow_enabled=dict(type='bool'),
+        implicit_flow_enabled=dict(type='bool'),
+        direct_access_grants_enabled=dict(type='bool'),
+        service_accounts_enabled=dict(type='bool'),
+        authorization_services_enabled=dict(type='bool'),
+        public_client=dict(type='bool'),
+        frontchannel_logout=dict(type='bool'),
+        protocol=dict(type='str', choices=['openid-connect', 'saml']),
+        attributes=dict(type='dict'),
+        full_scope_allowed=dict(type='bool'),
+        node_re_registration_timeout=dict(type='int'),
+        registered_nodes=dict(type='dict'),
+        client_template=dict(type='str'),
+        use_template_config=dict(type='bool'),
+        use_template_scope=dict(type='bool'),
+        use_template_mappers=dict(type='bool'),
+        protocol_mappers=dict(type='list'),
+        authorization_settings=dict(type='dict'),
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True,
+                           required_one_of=([['client_id', 'id']]))
+
+    result = dict(changed=False, msg='', diff={}, proposed={}, existing={}, end_state={})
+
+    # Obtain access token, initialize API
+    kc = KeycloakAPI(module)
+
+    realm = module.params.get('realm')
+    id = module.params.get('id')
```

Comment:
'id' is also a python built in, so it may be better to not shadow it here. In this code, it doesn't look like it would cause any problems however.

But, then that would also be inconsistent with the rest of the module args and REST api names.

So either way is ok with me, just pointing it out. I think if I was going to try to avoid it, I might add a dict or class that maps the local ansible module arg name to the remote rest api name, but that might be overkill, so not suggesting this as required change. But maybe something like:

``` python
api_map = {'id': 'id',
                    'direct_access_grants_enabled': 'directAccessGrantsEnabled',
                   # other names hre
}
```

so code could do something like:

``` python

if api_map['client_id'] not in updated_client:
    stuff()
```

Probably overkill, but also might provide some isolation against potential keycloak rest api name changes. Not worried about it now though so not a blocker for merge.

Reply:
Hrrm. A mapping like that might become useful, though for now I'd forego it for brevity.

id being shadowed bothers me as well (if only for the nice squiggly lines the editor does, but also in case somebody working on the module is surprised by it). I'll just rename the python variable itself to cid as a compromise; it doesn't change anything user-visible.